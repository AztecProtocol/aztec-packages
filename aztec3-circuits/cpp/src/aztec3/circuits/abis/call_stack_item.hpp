#pragma once
#include "function_data.hpp"
#include "private_circuit_public_inputs.hpp"
#include "public_circuit_public_inputs.hpp"

#include <barretenberg/stdlib/primitives/witness/witness.hpp>
#include <aztec3/utils/types/circuit_types.hpp>
#include <aztec3/utils/types/convert.hpp>
#include <aztec3/utils/types/native_types.hpp>

namespace aztec3::circuits::abis {

using aztec3::utils::types::CircuitTypes;
using aztec3::utils::types::NativeTypes;
using std::conditional;
using std::is_same;

enum class CallType {
    Public,
    Private,
};

template <typename NCT, CallType call_type> struct CallStackItem {
    typedef typename NCT::address address;
    typedef typename NCT::boolean boolean;
    typedef typename NCT::fr fr;

    template <typename T>
    using PublicInputs = typename std::
        conditional<call_type == CallType::Public, PublicCircuitPublicInputs<T>, PrivateCircuitPublicInputs<T>>::type;

    // This is the _actual_ contract address relating to where this function's code resides in the
    // contract tree. Regardless of whether this is a call or delegatecall, this
    // `contract_address` _does not change_. Amongst other things, it's used as a lookup for
    // getting the correct code from the tree. There is a separate `storage_contract_address`
    // within a CallStackItem which varies depending on whether this is a call or delegatecall.
    address contract_address = 0;
    FunctionData<NCT> function_data{};
    PublicInputs<NCT> public_inputs{};

    boolean operator==(CallContext<NCT> const& other) const
    {
        return contract_address == other.contract_address && function_data == other.function_data &&
               public_inputs == other.public_inputs;
    };

    template <typename Composer>
    CallStackItem<CircuitTypes<Composer>, call_type> to_circuit_type(Composer& composer) const
    {
        static_assert((std::is_same<NativeTypes, NCT>::value));

        // Capture the composer:
        auto to_ct = [&](auto& e) { return aztec3::utils::types::to_ct(composer, e); };

        CallStackItem<CircuitTypes<Composer>, call_type> call_stack_item = {
            to_ct(contract_address),
            function_data.to_circuit_type(composer),
            public_inputs.to_circuit_type(composer),
        };

        return call_stack_item;
    };

    fr hash() const
    {
        std::vector<fr> inputs = {
            contract_address.to_field(),
            function_data.hash(),
            public_inputs.hash(),
        };

        fr call_stack_item_hash = NCT::compress(inputs, GeneratorIndex::CALL_STACK_ITEM);

        return call_stack_item_hash;
    }
}; // namespace aztec3::circuits::abis

template <typename NCT, CallType call_type>
void read(uint8_t const*& it, CallStackItem<NCT, call_type>& call_stack_item)
{
    using serialize::read;

    read(it, call_stack_item.contract_address);
    read(it, call_stack_item.function_data);
    read(it, call_stack_item.public_inputs);
};

template <typename NCT, CallType call_type>
void write(std::vector<uint8_t>& buf, CallStackItem<NCT, call_type> const& call_stack_item)
{
    using serialize::write;

    write(buf, call_stack_item.contract_address);
    write(buf, call_stack_item.function_data);
    write(buf, call_stack_item.public_inputs);
};

template <typename NCT, CallType call_type>
std::ostream& operator<<(std::ostream& os, CallStackItem<NCT, call_type> const& call_stack_item)
{
    return os << "contract_address: " << call_stack_item.contract_address << "\n"
              << "function_data: " << call_stack_item.function_data << "\n"
              << "public_inputs: " << call_stack_item.public_inputs << "\n";
}

} // namespace aztec3::circuits::abis