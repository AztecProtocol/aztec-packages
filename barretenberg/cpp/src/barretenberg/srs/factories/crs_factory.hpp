#pragma once
#include "barretenberg/common/mem.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/g2.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include <cstddef>

namespace bb::pairing {
struct miller_lines;
} // namespace bb::pairing

namespace bb::srs::factories {

/**
 * A prover crs representation.
 */
template <typename Curve> class ProverCrs {
  public:
    virtual ~ProverCrs() = default;
    ;

    /**
     *  @brief Returns the monomial points in a form to be consumed by scalar_multiplication pippenger algorithm.
     */
    virtual typename Curve::AffineElement* get_monomial_points() = 0;
    virtual size_t get_monomial_size() const = 0;
};

template <typename Curve> class VerifierCrs {
  public:
    virtual ~VerifierCrs() = default;
};
template <> class VerifierCrs<curve::BN254> {
    using Curve = curve::BN254;

  public:
    virtual Curve::G2AffineElement get_g2x() const = 0;
    /**
     * @brief As the G_2 element of the CRS is fixed, we can precompute the operations performed on it during the
     * pairing algorithm to optimize pairing computations.
     */
    virtual bb::pairing::miller_lines const* get_precomputed_g2_lines() const = 0;
    /**
     *  @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

template <> class VerifierCrs<curve::Grumpkin> {
    using Curve = curve::Grumpkin;

  public:
    /**
     * @brief Returns the G_1 elements in the CRS after the pippenger point table has been applied on them
     *
     */
    virtual Curve::AffineElement* get_monomial_points() const = 0;
    virtual size_t get_monomial_size() const = 0;
    /**
     * @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

/**
 * A factory class to return the prover crs and verifier crs on request.
 * You can construct an empty placeholder factory, because composers need to be given a factory at construction time.
 */
template <typename Curve> class CrsFactory {
  public:
    CrsFactory() = default;
    CrsFactory(CrsFactory&& other) = default;
    virtual ~CrsFactory() = default;
    virtual std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> get_prover_crs(size_t) { return nullptr; }
    virtual std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> get_verifier_crs([[maybe_unused]] size_t degree = 0)
    {
        return nullptr;
    }
};

} // namespace bb::srs::factories