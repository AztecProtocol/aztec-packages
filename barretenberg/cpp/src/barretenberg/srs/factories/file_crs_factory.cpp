#include "file_crs_factory.hpp"
#include "../io.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/pairing.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include "barretenberg/ecc/scalar_multiplication/point_table.hpp"
#include "barretenberg/ecc/scalar_multiplication/scalar_multiplication.hpp"

namespace bb::srs::factories {

FileVerifierCrs<curve::BN254>::FileVerifierCrs(std::string const& path, const size_t)
    : precomputed_g2_lines((bb::pairing::miller_lines*)(aligned_alloc(64, sizeof(bb::pairing::miller_lines) * 2)))
{
    using Curve = curve::BN254;
    auto point_buf = scalar_multiplication::point_table_alloc<Curve::AffineElement>(1);
    srs::IO<Curve>::read_transcript_g1(point_buf.get(), 1, path);
    srs::IO<curve::BN254>::read_transcript_g2(g2_x, path);
    bb::pairing::precompute_miller_lines(bb::g2::one, precomputed_g2_lines[0]);
    bb::pairing::precompute_miller_lines(g2_x, precomputed_g2_lines[1]);
    first_g1 = point_buf[0];
}

FileVerifierCrs<curve::BN254>::~FileVerifierCrs()
{
    aligned_free(precomputed_g2_lines);
}

FileVerifierCrs<curve::Grumpkin>::FileVerifierCrs(std::string const& path, const size_t num_points)
    : num_points(num_points)
{
    using Curve = curve::Grumpkin;
    monomials_ = scalar_multiplication::point_table_alloc<Curve::AffineElement>(num_points);
    srs::IO<Curve>::read_transcript_g1(monomials_.get(), num_points, path);
    scalar_multiplication::generate_pippenger_point_table<Curve>(monomials_.get(), monomials_.get(), num_points);
    first_g1 = monomials_[0];
};

curve::Grumpkin::AffineElement* FileVerifierCrs<curve::Grumpkin>::get_monomial_points() const
{
    return monomials_.get();
}

size_t FileVerifierCrs<curve::Grumpkin>::get_monomial_size() const
{
    return num_points;
}

template <typename Curve>
FileCrsFactory<Curve>::FileCrsFactory(std::string path, size_t initial_degree)
    : path_(std::move(path))
    , degree_(initial_degree)
{}

template <typename Curve>
std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> FileCrsFactory<Curve>::get_prover_crs(size_t degree)
{
    if (degree != degree_ || !prover_crs_) {
        prover_crs_ = std::make_shared<FileProverCrs<Curve>>(degree, path_);
        degree_ = degree;
    }
    return prover_crs_;
}

template <typename Curve>
std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> FileCrsFactory<Curve>::get_verifier_crs(size_t degree)
{
    if (degree != degree_ || !verifier_crs_) {
        verifier_crs_ = std::make_shared<FileVerifierCrs<Curve>>(path_, degree);
        degree_ = degree;
    }
    return verifier_crs_;
}

template class FileProverCrs<curve::BN254>;
template class FileProverCrs<curve::Grumpkin>;
template class FileCrsFactory<curve::BN254>;
template class FileCrsFactory<curve::Grumpkin>;

} // namespace bb::srs::factories
