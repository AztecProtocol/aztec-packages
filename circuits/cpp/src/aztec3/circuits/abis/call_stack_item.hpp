#pragma once
#include "function_signature.hpp"
#include "private_circuit_public_inputs.hpp"
#include "public_circuit_public_inputs.hpp"

#include <stdlib/primitives/witness/witness.hpp>
#include <stdlib/types/circuit_types.hpp>
#include <stdlib/types/convert.hpp>
#include <stdlib/types/native_types.hpp>

namespace aztec3::circuits::abis {

using plonk::stdlib::witness_t;
using plonk::stdlib::types::CircuitTypes;
using plonk::stdlib::types::NativeTypes;
using std::conditional;
using std::is_same;

enum class CallType {
    Public,
    Private,
};

template <typename NCT, CallType call_type> struct CallStackItem {
    typedef typename NCT::address address;
    typedef typename NCT::boolean boolean;
    typedef typename NCT::fr fr;

    template <typename T>
    using PublicInputs = typename std::
        conditional<call_type == CallType::Public, PublicCircuitPublicInputs<T>, PrivateCircuitPublicInputs<T>>::type;

    address
        contract_address; // This is the _actual_ contract address relating to where this function's code resides in the
                          // contract tree. Regardless of whether this is a call or delegatecall, this
                          // `contract_address` _does not change_. Amongst other things, it's used as a lookup for
                          // getting the correct code from the tree. There is a separate `storage_contract_address`
                          // within a CallStackItem which varies depending on whether this is a call or delegatecall.
    FunctionSignature<NCT> function_signature;
    PublicInputs<NCT> public_inputs;

    bool operator==(CallStackItem<NCT, call_type> const&) const = default;

    template <typename T> static CallStackItem<NCT, call_type> empty()
    {
        return { 0, FunctionSignature<NCT>::empty(), PublicInputs<NCT>::empty() };
    };

    template <typename Composer>
    CallStackItem<CircuitTypes<Composer>, call_type> to_circuit_type(Composer& composer) const
    {
        static_assert((std::is_same<NativeTypes, NCT>::value));

        // Capture the composer:
        auto to_ct = [&](auto& e) { return plonk::stdlib::types::to_ct(composer, e); };

        CallStackItem<CircuitTypes<Composer>, call_type> call_stack_item = {
            to_ct(contract_address),
            function_signature.to_circuit_type(composer),
            public_inputs.to_circuit_type(composer),
        };

        return call_stack_item;
    };

    fr hash() const
    {
        std::vector<fr> inputs = {
            contract_address.to_field(),
            function_signature.hash(),
            public_inputs.hash(),
        };

        fr call_stack_item_hash = NCT::compress(inputs, GeneratorIndex::CALL_STACK_ITEM);

        return call_stack_item_hash;
    }
}; // namespace aztec3::circuits::abis

template <typename NCT, CallType call_type>
void read(uint8_t const*& it, CallStackItem<NCT, call_type>& call_stack_item)
{
    using serialize::read;

    read(it, call_stack_item.contract_address);
    read(it, call_stack_item.function_signature);
    read(it, call_stack_item.public_inputs_hash);
};

template <typename NCT, CallType call_type>
void write(std::vector<uint8_t>& buf, CallStackItem<NCT, call_type> const& call_stack_item)
{
    using serialize::write;

    write(buf, call_stack_item.contract_address);
    write(buf, call_stack_item.function_signature);
    write(buf, call_stack_item.public_inputs_hash);
};

template <typename NCT, CallType call_type>
std::ostream& operator<<(std::ostream& os, CallStackItem<NCT, call_type> const& call_stack_item)
{
    return os << "contract_address: " << call_stack_item.contract_address << "\n"
              << "function_signature: " << call_stack_item.function_signature << "\n"
              << "public_inputs: " << call_stack_item.public_inputs << "\n";
}

} // namespace aztec3::circuits::abis