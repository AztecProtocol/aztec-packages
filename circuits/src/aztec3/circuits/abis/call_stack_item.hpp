#pragma once
#include "function_signature.hpp"
#include "call_context.hpp"
#include "private_circuit_public_inputs.hpp"
#include "public_circuit_public_inputs.hpp"
#include <stdlib/primitives/witness/witness.hpp>
#include <stdlib/types/native_types.hpp>
#include <stdlib/types/circuit_types.hpp>
#include <stdlib/types/convert.hpp>

namespace aztec3::circuits::abis {

using plonk::stdlib::witness_t;
using plonk::stdlib::types::CircuitTypes;
using plonk::stdlib::types::NativeTypes;
using std::conditional;
using std::is_same;

enum class CallType {
    Public,
    Private,
};

template <typename NCT, CallType call_type> struct CallStackItem {
    typedef typename NCT::boolean boolean;
    typedef typename NCT::fr fr;

    template <typename T>
    using PublicInputs = typename std::
        conditional<call_type == CallType::Public, PublicCircuitPublicInputs<T>, PrivateCircuitPublicInputs<T>>::type;

    FunctionSignature<NCT> function_signature;
    PublicInputs<NCT> public_inputs; // TODO: can we just do args?
    CallContext<NCT> call_context;
    boolean is_delegate_call = false;
    boolean is_static_call = false;

    bool operator==(CallStackItem<NCT, call_type> const&) const = default;

    template <typename T> static CallStackItem<NCT, call_type> empty()
    {
        return { FunctionSignature<NCT>::empty(), PublicInputs<NCT>::empty(), CallContext<NCT>::empty(), 0, 0 };
    };

    template <typename Composer>
    CallStackItem<CircuitTypes<Composer>, call_type> to_circuit_type(Composer& composer) const
    {
        static_assert((std::is_same<NativeTypes, NCT>::value));

        // Capture the composer:
        auto to_ct = [&](auto& e) { return plonk::stdlib::types::to_ct(composer, e); };

        CallStackItem<CircuitTypes<Composer>, call_type> call_stack_item = {
            function_signature.to_circuit_type(composer),
            public_inputs.to_circuit_type(composer),
            call_context.to_circuit_type(composer),
            to_ct(is_delegate_call),
            to_ct(is_static_call),
        };

        return call_stack_item;
    };

    fr hash() const
    {
        std::vector<fr> inputs = {
            function_signature.hash(), public_inputs.hash(), call_context.hash(),
            fr(is_delegate_call),      fr(is_static_call),
        };

        return NCT::compress(inputs, GeneratorIndex::CALL_STACK_ITEM);
    }
};

template <typename NCT, CallType call_type>
void read(uint8_t const*& it, CallStackItem<NCT, call_type>& call_stack_item)
{
    using serialize::read;

    read(it, call_stack_item.function_signature);
    read(it, call_stack_item.public_inputs_hash);
    read(it, call_stack_item.call_context);
    read(it, call_stack_item.is_delegate_call);
    read(it, call_stack_item.is_callback);
};

template <typename NCT, CallType call_type>
void write(std::vector<uint8_t>& buf, CallStackItem<NCT, call_type> const& call_stack_item)
{
    using serialize::write;

    write(buf, call_stack_item.function_signature);
    write(buf, call_stack_item.public_inputs_hash);
    write(buf, call_stack_item.call_context);
    write(buf, call_stack_item.is_delegate_call);
    write(buf, call_stack_item.is_static_call);
};

template <typename NCT, CallType call_type>
std::ostream& operator<<(std::ostream& os, CallStackItem<NCT, call_type> const& call_stack_item)
{
    return os << "function_signature: " << call_stack_item.function_signature << "\n"
              << "public_inputs: " << call_stack_item.public_inputs << "\n"
              << "call_context: " << call_stack_item.call_context << "\n"
              << "is_delegate_call: " << call_stack_item.is_delegate_call << "\n"
              << "is_static_call: " << call_stack_item.is_static_call << "\n";
}

} // namespace aztec3::circuits::abis