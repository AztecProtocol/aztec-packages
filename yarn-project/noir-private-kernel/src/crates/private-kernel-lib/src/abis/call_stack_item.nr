use crate::address::Address;
use crate::abis::function_data::FunctionData;
use dep::aztec::constants_gen::{MAX_READ_REQUESTS_PER_CALL};
use crate::abis::private_circuit_public_inputs::PrivateCircuitPublicInputs;
use crate::abis::public_circuit_public_inputs::PublicCircuitPublicInputs;
use dep::aztec::constants_gen;

// TODO(Noir-bug): These should be type aliases

struct PublicCallStackItem {
    inner : CallStackItem<PublicCircuitPublicInputs>
}

struct PrivateCallStackItem {
    inner : CallStackItem<PrivateCircuitPublicInputs>
}

// Parametrize function over the public inputs.
// The public inputs can be from either a public or a private circuit
// TODO: We could monomorphize this struct and create a PrivateCallStackItem and 
// TODO: a PublicCallStackItem
struct CallStackItem<AppCircuitPublicInputs> {
    // Comment copied from the cpp codebase
    //
    // This is the _actual_ contract address relating to where this function's code resides in the
    // contract tree. Regardless of whether this is a call or delegatecall, this
    // `contract_address` _does not change_. Amongst other things, it's used as a lookup for
    // getting the correct code from the tree. There is a separate `storage_contract_address`
    // within a CallStackItem which varies depending on whether this is a call or delegatecall.
    contract_address : Address,

    public_inputs : AppCircuitPublicInputs,

    // True if this call stack item represents a request to execute a function rather than a
    // fulfilled execution. Used when enqueuing calls from private to public functions.
    is_execution_request : bool,

    function_data : FunctionData,
}

// TODO(Traits): Defining a hash trait bound
// Would allow the impl to be over `T : Hash`
// Where T is public inputs

// Noir(bug): If we can use a type alias instead of a new struct
// then we can remove the getter methods below for things like
// public_inputs and contract_address
impl PublicCallStackItem {
    fn hash(self) -> Field {
        dep::std::hash::pedersen_with_separator([
            self.inner.contract_address.to_field(),
            self.inner.function_data.hash(),
            self.inner.public_inputs.hash(),
        ], constants_gen::GENERATOR_INDEX__CALL_STACK_ITEM)[0]
    }

    fn public_inputs(self) -> PublicCircuitPublicInputs {
        self.inner.public_inputs
    }

    fn contract_address(self) -> Address{
        self.inner.contract_address
    }

    fn function_data(self) -> FunctionData{
        self.inner.function_data
    }
}
impl PrivateCallStackItem {
    fn hash(self) -> Field {
        dep::std::hash::pedersen_with_separator([
            self.inner.contract_address.to_field(),
            self.inner.function_data.hash(),
            self.inner.public_inputs.hash(),
        ], constants_gen::GENERATOR_INDEX__CALL_STACK_ITEM)[0]
    }

    fn public_inputs(self) -> PrivateCircuitPublicInputs {
        self.inner.public_inputs
    }

    fn contract_address(self) -> Address{
        self.inner.contract_address
    }

    fn function_data(self) -> FunctionData{
        self.inner.function_data
    }
}
