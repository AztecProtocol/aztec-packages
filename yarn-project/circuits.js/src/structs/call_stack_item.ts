import { AztecAddress } from '@aztec/foundation/aztec-address';

import { computePublicCallStackItemHash } from '../abis/abis.js';
import { CircuitsWasm } from '../index.js';
import { serializeToBuffer } from '../utils/serialize.js';
import { FunctionData } from './function_data.js';
import { PrivateCircuitPublicInputs } from './private_circuit_public_inputs.js';
import { PublicCircuitPublicInputs } from './public_circuit_public_inputs.js';

/**
 * Call stack item on a private call.
 * @see cpp/src/aztec3/circuits/abis/call_stack_item.hpp.
 */
export class PrivateCallStackItem {
  constructor(
    /**
     * Address of the contract on which the function is invoked.
     */
    public contractAddress: AztecAddress,
    /**
     * Data identifying the function being called.
     */
    public functionData: FunctionData,
    /**
     * Public inputs to the private kernel circuit.
     */
    public publicInputs: PrivateCircuitPublicInputs,
    /**
     * Whether the current callstack item should be considered a public fn execution request.
     */
    public readonly isExecutionRequest: boolean,
  ) {
    if (isExecutionRequest) {
      throw new Error('boolean isExecutionRequest must be set to true for a PrivateCallStackItem object');
    }
  }

  toBuffer() {
    return serializeToBuffer(this.contractAddress, this.functionData, this.publicInputs, this.isExecutionRequest);
  }

  /**
   * Returns a new instance of PrivateCallStackItem with zero contract address, function data and public inputs.
   * @returns A new instance of PrivateCallStackItem with zero contract address, function data and public inputs.
   */
  public static empty(): PrivateCallStackItem {
    return new PrivateCallStackItem(
      AztecAddress.ZERO,
      FunctionData.empty({ isPrivate: true }),
      PrivateCircuitPublicInputs.empty(),
      false,
    );
  }
}

/**
 * Call stack item on a public call.
 * @see cpp/src/aztec3/circuits/abis/call_stack_item.hpp.
 */
export class PublicCallStackItem {
  constructor(
    /**
     * Address of the contract on which the function is invoked.
     */
    public contractAddress: AztecAddress,
    /**
     * Data identifying the function being called.
     */
    public functionData: FunctionData,
    /**
     * Public inputs to the public kernel circuit.
     */
    public publicInputs: PublicCircuitPublicInputs,
    /**
     * Whether the current callstack item should be considered a public fn execution request.
     */
    public isExecutionRequest: boolean,
  ) {}

  toBuffer() {
    return serializeToBuffer(this.contractAddress, this.functionData, this.publicInputs, this.isExecutionRequest);
  }

  /**
   * Returns a new instance of PublicCallStackItem with zero contract address, function data and public inputs.
   * @returns A new instance of PublicCallStackItem with zero contract address, function data and public inputs.
   */
  public static empty(): PublicCallStackItem {
    return new PublicCallStackItem(
      AztecAddress.ZERO,
      FunctionData.empty({ isPrivate: false }),
      PublicCircuitPublicInputs.empty(),
      false,
    );
  }

  isEmpty() {
    return this.contractAddress.isZero() && this.functionData.isEmpty() && this.publicInputs.isEmpty();
  }

  /**
   * Computes this call stack item hash.
   * @returns Hash.
   */
  public async hash() {
    return computePublicCallStackItemHash(await CircuitsWasm.get(), this);
  }
}
