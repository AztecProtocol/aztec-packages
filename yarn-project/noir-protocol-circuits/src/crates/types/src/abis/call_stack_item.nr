use crate::abis::{
    function_data::FunctionData,
    private_circuit_public_inputs::PrivateCircuitPublicInputs,
    public_circuit_public_inputs::PublicCircuitPublicInputs,
};
use crate::address::AztecAddress;
use crate::constants::{
    GENERATOR_INDEX__CALL_STACK_ITEM,
};

struct PrivateCallStackItem {
    // This is the _actual_ contract address relating to where this function's code resides in the
    // contract tree. Regardless of whether this is a call or delegatecall, this
    // `contract_address` _does not change_. Amongst other things, it's used as a lookup for
    // getting the correct code from the tree. There is a separate `storage_contract_address`
    // within a CallStackItem which varies depending on whether this is a call or delegatecall.
    contract_address: AztecAddress,
    public_inputs: PrivateCircuitPublicInputs,
    function_data: FunctionData,
    // Not really needed for PrivateCallStackItem.
    is_execution_request: bool,
}

impl PrivateCallStackItem {
    fn hash(self) -> Field {
        dep::std::hash::pedersen_hash_with_separator([
            self.contract_address.to_field(),
            self.function_data.hash(),
            self.public_inputs.hash(),
        ], GENERATOR_INDEX__CALL_STACK_ITEM)
    }
}

struct PublicCallStackItem {
    contract_address: AztecAddress,
    public_inputs: PublicCircuitPublicInputs,
    function_data: FunctionData,
    // True if this call stack item represents a request to execute a function rather than a
    // fulfilled execution. Used when enqueuing calls from private to public functions.
    is_execution_request: bool,
}

impl PublicCallStackItem {
    fn hash(self) -> Field {
        let item = if self.is_execution_request {
            self.as_execution_request()
        } else {
            self
        };

        dep::std::hash::pedersen_hash_with_separator([
            item.contract_address.to_field(),
            item.function_data.hash(),
            item.public_inputs.hash(),
        ], GENERATOR_INDEX__CALL_STACK_ITEM)
    }

    fn as_execution_request(self) -> Self {
        let public_inputs = self.public_inputs;
        let mut request_public_inputs: PublicCircuitPublicInputs = dep::std::unsafe::zeroed();
        request_public_inputs.call_context = public_inputs.call_context;
        request_public_inputs.args_hash = public_inputs.args_hash;

        PublicCallStackItem {
            contract_address: self.contract_address,
            function_data: self.function_data,
            is_execution_request: true,
            public_inputs: request_public_inputs,
        }
    }
}
