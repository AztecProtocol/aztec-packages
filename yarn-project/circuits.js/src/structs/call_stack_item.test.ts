import { makePrivateCallStackItem } from '../tests/factories.js';
import { PrivateCallStackItem } from './call_stack_item.js';

describe('PrivateCallStackItem', () => {
  it('serializes to buffer and deserializes it back', () => {
    const randomInt = Math.floor(Math.random() * 1000);
    const expected = makePrivateCallStackItem(randomInt);
    const buffer = expected.toBuffer();
    const res = PrivateCallStackItem.fromBuffer(buffer);
    expect(res).toEqual(expected);
  });

  it('serializes to field array and deserializes it back', () => {
    const randomInt = Math.floor(Math.random() * 1000);
    const expected = makePrivateCallStackItem(randomInt);

    const fieldArray = expected.toFields();
    const res = PrivateCallStackItem.fromFields(fieldArray);
    expect(res).toEqual(expected);
  });

  it('computes hash', () => {
    const seed = 9870243;
    const PrivateCallStackItem = makePrivateCallStackItem(seed);
    const hash = PrivateCallStackItem.hash();
    expect(hash).toMatchSnapshot();
  });
});
