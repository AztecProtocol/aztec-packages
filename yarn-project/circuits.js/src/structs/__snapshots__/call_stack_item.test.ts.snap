// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateCallStackItem computes hash 1`] = `
Fr {
  "asBigInt": 18753511695134949302571620808052147141303876532454748772518515365970855750244n,
  "asBuffer": {
    "data": [
      41,
      118,
      28,
      137,
      68,
      13,
      118,
      183,
      92,
      239,
      142,
      84,
      234,
      235,
      13,
      55,
      245,
      66,
      184,
      218,
      252,
      190,
      14,
      252,
      58,
      237,
      34,
      161,
      161,
      6,
      106,
      100,
    ],
    "type": "Buffer",
  },
}
`;
